### PR TITLE
What's Yours Is Mined

### PR DESCRIPTION
**Pull request by mistyLuminescence**
TL;DR:
- Mines are now dangerous. Don't step on them.
- Fragmentation mines added. Don't step on them either.
- You can now shoot or explode mines to detonate them from afar (50% chance).
- You can now disarm mines with hacking tools. If you hear beeping, run. Fast.
- It is possible to commit no mistakes and still lose. That is not a weakness; that is life. Stay alive, run fast.
- When you disarm a mine, you can pick it up and move it somewhere else and reactivate it.
- You know mines you disarmed and moved and then rearmed? Don't step on them.
- How to disarm a mine: pulse until it pings, then cut. Run if it beeps. Don't cut blind.
> 

**ENGINEER FIELD MANUAL 5-69**
Take your **screwdriver** and carefully open the panel of the mine. Without opening the panel, the mine cannot be disarmed.
Take your **multitool** and select a wire; note that each mine has a **different wire configuration**, so what works on one might not work on another.
Pulse the first wire with your **multitool** (recommended *f*: 536 MHz). One of three things will happen:

- A beep: this wire is either the detonation wire or a dummy wire (50/50). **Do not cut this wire**. Move on.
- A **series of rapid beeping**: this wire is the timed detonation wire. You have **two seconds to run**.
- A **ping**: this wire is the disarm wire, dummy wire, or trap wire. Cut this wire with your **wirecutters**.

After cutting a pinging wire, one of three things will happen:
- Nothing. This was the dummy wire. Move onto the next wire.
- A **series of rapid beeping**: this wire is the trap wire. You have **two seconds to run**.
- A **click**. The mine is disarmed, and can safely be moved.

Do **not** attempt to disarm a mine without a multitool. Doing so will likely result in loss of life and/or limb, and will forfeit your guaranteed medical care with NanoTrasen.

In the absence of proper equipment, a land mine may be remotely detonated via ballistic or energy impact (including hunting rifles) or secondary explosive charge (e.g. grenade, another mine).

The most common types of mines you will encounter are high explosive and fragmentation.
High explosive mines are marked with 'HE' and a grenade symbol. They explode violently when stepped on or detonated incorrectly. 
Fragmentation mines are marked with 'FRAG' and a grenade symbol. They explode into a shower of shrapnel when stepped on or detonated incorrectly.

Stepping on an active land mine must be avoided at all costs; the detonation will be much more powerful when directly above the mine than even one meter away. This may be exploited by crafty survivalists to lure hostile wildlife into the land mine's explosive radius.

To arm a disabled land mine, press the large red button in the center of the mine, beginning the ten-second placement timer. Ensure that the mine is correctly placed within ten seconds of activation. Do not tamper with the mine once placed.

### CODE REVIEW

**Comment by Anewbe** [comment]
Redo your opening post and/or title to say specifically what this PR does. Bullet points are acceptable.

File: code/game/objects/effects/mines.dm [hunk 1]
```diff
@@ -68,38 +127,158 @@
...
+	triggerproc = "triggerfrag"
+	mineitemtype = /obj/item/weapon/mine/frag
+	var/fragment_types = list(/obj/item/projectile/bullet/pellet/fragment)
+	var/num_fragments = 20  //total number of fragments produced by the grenade
+	//The radius of the circle used to launch projectiles. Lower values mean less projectiles are used but if set too low gaps may appear in the spread pattern
+	var/spread_range = 7
+
+/obj/item/weapon/mine
+	name = "mine"
+	desc = "Normal procedure is to jump 200 feet in the air and scatter yourself over a wide area."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 2]
```diff
@@ -1,32 +1,89 @@
 /obj/effect/mine
-	name = "Mine"
-	desc = "I Better stay away from that thing."
-	density = 1
+	name = "mine"
+	desc = "Normal procedure is to jump 200 feet in the air and scatter yourself over a wide area."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 3]
```diff
@@ -68,38 +127,158 @@
...
+	if(!O)
+		return
+	src.fragmentate(O, 20, 7, list(/obj/item/projectile/bullet/pellet/fragment)) //only 20 weak fragments because you're stepping directly on it
+	qdel(src)
+
+
 /obj/effect/mine/dnascramble
-	name = "Radiation Mine"
+	name = "radiation mine"
+	desc = "Through the rain of Strontium 90..."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 4]
```diff
@@ -68,38 +127,158 @@
...
 	icon_state = "uglymine"
 	triggerproc = "triggerrad"
+	mineitemtype = /obj/item/weapon/mine/dnascramble
 
 /obj/effect/mine/phoron
-	name = "Phoron Mine"
+	name = "incendiary mine"
+	desc = "We didn't start the fire. This mine did."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 5]
```diff
@@ -68,38 +127,158 @@
...
 	icon_state = "uglymine"
 	triggerproc = "triggerphoron"
+	mineitemtype = /obj/item/weapon/mine/phoron
 
 /obj/effect/mine/kick
-	name = "Kick Mine"
+	name = "kick mine"
+	desc = "Concentrated war crimes. Handle with care."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: This description is fine, because these are never going to be mapped in.

File: code/game/objects/effects/mines.dm [hunk 6]
```diff
@@ -68,38 +127,158 @@
...
-	name = "Stun Mine"
+	name = "stun mine"
+	desc = "Stronger than a Pan-Galactic, but only half as volatile."
 	icon_state = "uglymine"
 	triggerproc = "triggerstun"
+	mineitemtype = /obj/item/weapon/mine/stun
+
+/obj/effect/mine/frag
+	name = "fragmentation mine"
+	desc = "Ever wanted to become a pincushion?"
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 7]
```diff
@@ -68,38 +127,158 @@
...
+
+/obj/item/weapon/mine/proc/prime(mob/user as mob)
+	visible_message("\The [src.name] beeps as the priming sequence completes.")
+	new minetype(get_turf(src))
+	spawn(0)
+		qdel(src)
+
+/obj/item/weapon/mine/dnascramble
+	name = "radiation mine"
+	desc = "Through the rain of Strontium 90..."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 8]
```diff
@@ -68,38 +127,158 @@
...
+	icon_state = "uglymine"
+	minetype = /obj/effect/mine/dnascramble
+
+/obj/item/weapon/mine/phoron
+	name = "incendiary mine"
+	desc = "We didn't start the fire. This mine did."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 9]
```diff
@@ -68,38 +127,158 @@
...
+
+/obj/item/weapon/mine/kick
+	name = "kick mine"
+	desc = "Concentrated war crimes. Handle with care."
+	icon_state = "uglymine"
+	minetype = /obj/effect/mine/kick
+
+/obj/item/weapon/mine/n2o
+	name = "nitrous oxide mine"
+	desc = "Nitrous oxide is no laughing matter."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 10]
```diff
@@ -68,38 +127,158 @@
...
+	icon_state = "uglymine"
+	minetype = /obj/effect/mine/n2o
+
+/obj/item/weapon/mine/stun
+	name = "stun mine"
+	desc = "Stronger than a Pan-Galactic, but only half as volatile."
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 11]
```diff
@@ -68,38 +127,158 @@
...
+	icon_state = "uglymine"
+	minetype = /obj/effect/mine/stun
+
+/obj/item/weapon/mine/frag
+	name = "fragmentation mine"
+	desc = "Ever wanted to become a pincushion?"
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Let's get a more serious description, please.

File: code/game/objects/effects/mines.dm [hunk 12]
```diff
@@ -68,38 +127,158 @@
...
+			prime()
+			if(user)
+				msg_admin_attack("[user.name] ([user.ckey]) primed \a [src] (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[user.x];Y=[user.y];Z=[user.z]'>JMP</a>)")
+			return
+
+/obj/item/weapon/mine/proc/prime(mob/user as mob)
+	visible_message("\The [src.name] beeps as the priming sequence completes.")
+	new minetype(get_turf(src))
+	spawn(0)
+		qdel(src)
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Move the fingerprints to the new mine, please.

**Comment by mistyLuminescence** [comment]
there was a bullet pointed tl;dr summary already, moved it to the top and added a quick how-to on the new features

File: code/game/objects/effects/mines.dm [hunk 13]
```diff
@@ -1,32 +1,89 @@
...
+			to_chat(user, "<span class='danger'>Your tools brush against the activator! Run!</span>")
+			spawn(10)
+				call(src,triggerproc)(user)
+				return
+		if(prob(80) && isolated == 1)
+			to_chat(user, "<span class='notice'>Snip!</span>")
+			visible_message("<span class='danger'>[user] safely disarms \the [src]!</span>")
+			new mineitemtype(get_turf(src))
+			spawn(0)
+				qdel(src)
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: Move the fingerprints to the new mine, please.

**Comment by mistyLuminescence** [comment]
wire minigame is gonna take a bit, and the frag mine is way weaker than a grenade, but keep in mind that the chances of being in immediate proximity when it detonates is like... 15 times as high
step on it and you're in crit; fail to disarm it and you probably only get fractures and shrapnel (still bad. don't do this.)

can change descriptions 👍

**Comment by mistyLuminescence** [comment]
Wire minigame added, descriptions changed, original post updated.

**Comment by mistyLuminescence** [comment]
@Anewbe Changes made, thoughts?

File: code/game/objects/effects/mines.dm [hunk 14]
```diff
@@ -19,87 +40,205 @@
...
+/obj/effect/mine/attackby(obj/item/W as obj, mob/living/user as mob)
+	if(isscrewdriver(W))
+		panel_open = !panel_open
+		user.visible_message("<span class='warning'>[user] very carefully screws the mine's panel [panel_open ? "open" : "closed"].</span>",
+		"<span class='notice'>You very carefully screw the mine's panel [panel_open ? "open" : "closed"].</span>")
+		playsound(src.loc, W.usesound, 50, 1)
+
+	else if((iswirecutter(W) || ismultitool(W)) && panel_open)
+		interact(user)
+
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: There should probably be an `else ..()` in here somewhere.

File: code/game/objects/effects/mines.dm [hunk 15]
```diff
@@ -19,87 +40,205 @@
...
+			C.throw_mode_on()
+	else
+		to_chat(user, "You cancel \the [name]'s priming sequence.")
+		arming = 0
+		countdown = initial(countdown)
+		icon_state = initial(icon_state)
+		add_fingerprint(user)
+	return
+
+/obj/item/weapon/mine/proc/activate(mob/user as mob)
```
Comment (Anewbe) [inline code/game/objects/effects/mines.dm]: This whole proc could probably be done better as a do_after() in the attack_self().

**Comment by Anewbe** [comment]
Still needs a changelog.

File: code/datums/wires/mines.dm [hunk 16]
```diff
@@ -0,0 +1,81 @@
+/datum/wires/mines
+	wire_count = 6
+	random = 1
+	holder_type = /obj/effect/mine
+
+var/const/WIRE_DETONATE  = 1
```
Comment (Atermonera) [inline code/datums/wires/mines.dm]: Can you turn these into defines? It'll function identically, and just be handled by the compiler rather than having 6 more global vars floating around at runtime (Globals are generally bad unless they're exactly what you need for a certain situation)
